Public/Src/Sandbox/Windows/DetoursServices/DetouredFunctions.cpp

```diff
@@ -1778,6 +1778,224 @@ NTSTATUS NTAPI Detoured_ZwSetInformationFile(
+    {
+        command = wstring();
+        commandArgs = wstring();
+        return;
+    }
+
+    if (fullCommandLine[0] == L'"')
+    {
+        // Find the close quote. Might not be present which means the command
+        // is the full command line minus the initial quote.
+        size_t closeQuoteIndex = fullCommandLine.find('"', 1);
+        if (closeQuoteIndex >= 1)
+        {
+            if (closeQuoteIndex == fullCommandLine.length() - 1)
+            {
+                // Quotes cover entire command line.
+                command = fullCommandLine.substr(1, fullCommandLine.length() - 2);
+                trim_inplace(command);
+                commandArgs = wstring();
+            }
+            else
+            {
+                wstring noQuoteCommand = fullCommandLine.substr(1, closeQuoteIndex - 1);
+
+                // Find the next delimiting space after the close double-quote.
+                // For example a command line like "c:\program files"\foo we need to
+                // keep \foo and cut the quotes to produce c:\program files\foo
+                size_t spaceDelimiterIndex = fullCommandLine.find(L' ', closeQuoteIndex + 1);
+                if (spaceDelimiterIndex == wstring::npos)
+                {
+                    // No space, take everything through the end of the command line.
+                    spaceDelimiterIndex = fullCommandLine.length();
+                }
+
+                command = (noQuoteCommand +
+                    fullCommandLine.substr(closeQuoteIndex + 1, spaceDelimiterIndex - closeQuoteIndex));
+                trim_inplace(command);
+                commandArgs = fullCommandLine.substr(spaceDelimiterIndex + 1);
+                trim_inplace(commandArgs);
+            }
+        }
+        else
+        {
+            // No close quote. Take everything through the end of the command line as the command.
+            command = fullCommandLine.substr(1);
+            commandArgs = wstring();
+        }
+    }
+    else
+    {
+        // No open quote, pure space delimiter.
+        size_t spaceDelimiterIndex = fullCommandLine.find(' ');
+        if (spaceDelimiterIndex == wstring::npos)
+        {
+            // No space, take everything through the end of the command line.
+            spaceDelimiterIndex = fullCommandLine.length();
+        }
+
+        command = fullCommandLine.substr(0, spaceDelimiterIndex);
+        commandArgs = fullCommandLine.substr(spaceDelimiterIndex + 1);
+        trim_inplace(commandArgs);
+    }
+}
+
+static bool CommandArgsContainMatch(const wchar_t *commandArgs, const wchar_t *argMatch)
+{
+    if (argMatch == nullptr)
+    {
+        // No optional match, meaning always match.
+        return true;
+    }
+
+    return wcsstr(commandArgs, argMatch) != nullptr;
+}
+
+static bool ShouldShim(const wstring &command, const wchar_t *commandArgs)
+{
+    // Easy cases.
+    if (g_pShimProcessMatches == nullptr || g_pShimProcessMatches->empty())

```

Comment:
>himProcessMatches == nullptr || g_pShimProcessMatch [](start = 12, length = 51)

Shouldn't this be an assert case since this method only gets called when `g_substituteProcessExecutionShimPath` is not null. Why would the other side set g_substituteProcessExecutionShimPath and not the others, that should be an error case there too

Reply:
Ah you are overloading null with wildcards.

---
In reply to: [282960153](https://github.com/microsoft/BuildXL/pull/279#discussion_r282960153) [](ancestors = 282960153)

Reply:
No, this is still weird. 
If nothing is to be shimmed, it shouldn't set `substituteProcessExecutionShimPath`
if everything is to be shimmed: `g_ProcessExecutionShimAllProcesses` should be set to true.

Else it should be an error if there are zero matches.


---
In reply to: [282960313](https://github.com/microsoft/BuildXL/pull/279#discussion_r282960313) [](ancestors = 282960313,282960153)

Reply:
Added assert